pkg/util/inplaceupdate/inplace_update_vertical.go

```diff
@@ -0,0 +1,252 @@
+*/
+
+package inplaceupdate
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/appscode/jsonpatch"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	appspub "github.com/openkruise/kruise/apis/apps/pub"
+)
+
+// interface for In-place workload vertical scaling
+type VerticalUpdateInterface interface {
+	// ValidateResourcePatch validates and applies the resource patch to the UpdateSpec.
+	ValidateResourcePatch(op *jsonpatch.Operation, oldTemp *v1.PodTemplateSpec, updateSpec *UpdateSpec) error
+
+	// interface for container level vertical scaling api, such as k8s 1.27+
+
+	// SyncContainerResource Get the expected resource values of the container and its current status
+	SyncContainerResource(container *v1.ContainerStatus, state *appspub.InPlaceUpdateState)
+	// UpdateContainerResource Pass in the container to be modified and the expected resource values.
+	UpdateContainerResource(container *v1.Container, resource *v1.ResourceRequirements)
+	// IsContainerUpdateCompleted To determine whether the container has been successfully vertical updated
+	IsContainerUpdateCompleted(pod *v1.Pod, container *v1.Container, containerStatus *v1.ContainerStatus, lastContainerStatus appspub.InPlaceUpdateContainerStatus) bool
+
+	// interface for pod level vertical scaling api
+
+	// SyncPodResource Get the expected resource values of all containers in the pod and their current status
+	SyncPodResource(pod *v1.Pod, state *appspub.InPlaceUpdateState)
+	// UpdatePodResource All containers of a pod can be updated at once within this interface.
+	UpdatePodResource(pod *v1.Pod)
+	// IsPodUpdateCompleted To determine whether the pod has been successfully vertical updated
+	IsPodUpdateCompleted(pod *v1.Pod) bool
+}
+
+var verticalUpdateOperator VerticalUpdateInterface = nil
+
+// To register vertical update operations,
+// you can register different vertical update implementations here
+func init() {
+	// Now, we assume that there is a single standard per cluster, so register in init()
+	// TODO(Abner-1): Perhaps we should dynamically select the verticalUpdateOperator based on the pod metadata being processed.
+	// give us more suggestions if you need
+	if verticalUpdateOperator == nil {
+		verticalUpdateOperator = &VerticalUpdate{}
+	}
+}
+
+// VerticalUpdate represents the vertical scaling of k8s standard
+type VerticalUpdate struct{}
+
+var _ VerticalUpdateInterface = &VerticalUpdate{}
+
+func (v *VerticalUpdate) ValidateResourcePatch(op *jsonpatch.Operation, oldTemp *v1.PodTemplateSpec, updateSpec *UpdateSpec) error {
+	// for example: /spec/containers/0/resources/limits/cpu
+	words := strings.Split(op.Path, "/")
+	if len(words) != 7 {
+		return fmt.Errorf("invalid resource path: %s", op.Path)
+	}
+	idx, err := strconv.Atoi(words[3])
+	if err != nil || len(oldTemp.Spec.Containers) <= idx {
+		return fmt.Errorf("invalid container index: %s", op.Path)
+	}
+	quantity, err := resource.ParseQuantity(op.Value.(string))
```

Comment:
how about the following patch
```
  { "op": "remove", "path": "/spec/containers/0/resources/limits/cpu" }
```

it is possible to unset the container resource for one of the container in the pod of burstable qos class

Reply:
I will test this case.

Reply:
There is a releated issue in kubernetes https://github.com/kubernetes/kubernetes/pull/128718.